packages/webamp/js/components/VisPainter.ts

```diff
@@ -0,0 +1,748 @@
+      saData2[x] *= peakFalloff;
+      // Possible peak fall off values are
+      // 1.05f, 1.1f, 1.2f, 1.4f, 1.6f
+      // 1.1f is the default of a fresh new Winamp installation
+      if (saPeaks[x] <= 0) {
+        saPeaks[x] = 0;
+      }
+
+      if (smallVis) {
+        // SORRY NOTHING
+        // ironically enough the peaks do appear at the bottom here
+      } else {
+        if (Math.round(barPeak[x]) < 1) {
+          barPeak[x] = -3; // Push peaks outside the viewable area, this isn't a Modern Skin!
+        }
+      }
+
+      // skip rendering if x is 4
+      if (!(x == chunk + 3)) {
+        this.paintBar(
+          ctx,
+          x,
+          x,
+          Math.round(saFalloff[x]) - pushDown,
+          barPeak[x] + 1 - pushDown
+        );
+      }
+    }
+  }
+
+  /**
+   * 🟥
+   * 🟧🟧
+   * 🟨🟨🟨
+   * 🟩🟩🟩🟩
+   */
+  paintBarNormal(
+    ctx: CanvasRenderingContext2D,
+    x: number,
+    x2: number,
+    barHeight: number,
+    peakHeight: number
+  ) {
+    const h = ctx.canvas.height;
+    const y = h - barHeight;
+
+    ctx.drawImage(this._bar, 0, y, 1, h - y, x, y, x2 - x + 1, h - y);
+
+    if (this._vis.peaks) {
+      const peakY = h - peakHeight;
+      ctx.drawImage(this._peak, 0, 0, 1, 1, x, peakY, x2 - x + 1, 1);
+    }
+  }
+
+  /**
+   * 🟥
+   * 🟧🟥
+   * 🟨🟧🟥
+   * 🟩🟨🟧🟥
+   */
+  paintBarFire(
+    ctx: CanvasRenderingContext2D,
+    x: number,
+    x2: number,
+    barHeight: number,
+    peakHeight: number
+  ) {
+    const h = ctx.canvas.height;
+    let y = h - barHeight;
+
+    ctx.drawImage(
+      this._bar,
+      0,
+      0,
+      this._bar.width,
+      h - y,
+      x,
+      y,
+      x2 - x + 1,
+      h - y
+    );
+
+    if (this._vis.peaks) {
+      const peakY = h - peakHeight;
+      ctx.drawImage(this._peak, 0, 0, 1, 1, x, peakY, x2 - x + 1, 1);
+    }
+  }
+
+  /**
+   * 🟥
+   * 🟥🟧
+   * 🟥🟧🟨
+   * 🟥🟧🟨🟩
+   */
+  paintBarLine(
+    ctx: CanvasRenderingContext2D,
+    x: number,
+    x2: number,
+    barHeight: number,
+    peakHeight: number
+  ) {
+    const h = ctx.canvas.height;
+    let y = h - barHeight;
+
+    if (!this.logged) {
+      console.log("FIXME: Line drawing is currently Fire mode!");
+      this.logged = true;
+    }
+
+    ctx.drawImage(
+      this._bar,
+      0, // sx
+      0, // sy
+      this._bar.width, // sw
+      h - y, // sh
+      x,
+      y, //  dx,dy
+      x2 - x + 1, //dw
+      h - y //dh
+    );
+
+    if (this._vis.peaks) {
+      const peakY = h - peakHeight;
+      ctx.drawImage(this._peak, 0, 0, 1, 1, x, peakY, x2 - x + 1, 1);
+    }
+  }
+}
+
+//? =============================== OSCILLOSCOPE PAINTER ===============================
+
+type PaintWavFunction = (x: number, y: number) => void;
+
+function slice1st(
+  dataArray: Uint8Array,
+  sliceWidth: number,
+  sliceNumber: number
+): number {
+  const start = sliceWidth * sliceNumber;
+  return dataArray[start];
+}
+
+export class WavePaintHandler extends VisPaintHandler {
```

Comment:
Each of these painters is pretty complicated. Maybe we could put each one in their own file?